foundations/src/telemetry/tracing/live/event_output.rs

```diff
@@ -0,0 +1,94 @@
+//! Outputs telemetry spans in Chrome JSON trace format (i.e. the format used by about:tracing).
+use super::SharedSpanHandle;
+use std::time::SystemTime;
+
+/// Outputs a slice of shared spans as a Chrome JSON trace log.
+pub(crate) fn spans_to_trace_events(epoch: SystemTime, spans: &[SharedSpanHandle]) -> String {
+    use cf_rustracing::span::InspectableSpan;
+
+    let mut log_builder = TraceLogBuilder::new();
+
+    let end_timestamp = epoch
+        .elapsed()
+        .ok()
+        .and_then(|x| u64::try_from(x.as_nanos()).ok())
+        .unwrap_or(u64::MAX);
+
+    for span in spans {
+        let span_ref = span.read();
+        let span_state = span_ref.context().unwrap().state();
+        let trace_id = span_state.trace_id().to_string();
+        let name = span_ref.operation_name();
+
+        let start_ts = span_ref
+            .start_time()
+            .duration_since(epoch)
+            .ok()
+            .and_then(|x| u64::try_from(x.as_nanos()).ok())
+            .unwrap_or_default();
+
+        let end_ts = span_ref
+            .finish_time()
+            .and_then(|x| x.duration_since(epoch).ok())
+            .and_then(|x| u64::try_from(x.as_nanos()).ok())
+            .unwrap_or(end_timestamp);
+
+        log_builder.write_event(&trace_id, name, "", TraceEventType::Begin, start_ts);
+        log_builder.write_event(&trace_id, name, "", TraceEventType::End, end_ts);
+    }
+
+    log_builder.finalize(end_timestamp)
+}
+
+#[derive(Copy, Clone)]
+enum TraceEventType {
+    Begin,
+    End,
+}
+
+fn escape(s: &str) -> String {
+    s.escape_default().to_string()
+}
+
+struct TraceLogBuilder {
+    out: String,
+}
+
+impl TraceLogBuilder {
+    fn new() -> Self {
+        TraceLogBuilder {
+            out: "[".to_string(),
+        }
+    }
+
+    fn write_event(
+        &mut self,
+        trace_id: &str,
+        name: &str,
+        category: &str,
+        event_type: TraceEventType,
+        timestamp_us: u64,
+    ) {
+        self.out.push_str(&format!(
+            "{{\"pid\":1,\"name\":\"{}\",\"cat\":\"{}\",\"ph\":\"{}\",\"ts\":{},\"id\":\"{}\"}},",
```

Comment:
It would be nice to also emit the trace tags as `args`, but I understand if you don't have time to implement this right now.

Reply:
Tags are typed, so we'd need to generate JSON objects with serde or something rather than this quick and dirty approach.

Reply:
Yeah, let's leave that for a future iteration